File: Cpp_src/medium/palindromic_substrings_647.cc
#include <iostream>
using namespace std;

// Use Expand Around Center
// Time Complexity: O(n^2)
// Space Complexity:O(1)
// Runtime: 12 ms, faster than 58.78%
// Memory Usage: 15.5 MB, less than 16.00%
class Solution
{
public:
    int result = 0;
    int countSubstrings(string s)
    {
        for (int i = 0; i < s.size(); i++)
        {
            helper(s, i, i);
            helper(s, i, i + 1);
        }

        return result;
    }

    void helper(string s, int index_left, int index_right)
    {
        while (index_left >= 0 && index_right < s.size() && s[index_left] == s[index_right])
        {
            index_left--;
            index_right++;
            result++;
        }
    }
};

File: Golang_src/medium/palindromicSubstrings_647.go
package medium

// Use Expand Around Center:
// Time Complexity: O(n^2)
// Space Complexity:O(1)
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 2 MB, less than 52.52%
func countSubstrings(s string) int {
	count := 0
	strS := []rune(s)
	size := len(strS)

	for i := 0; i < size; i++ {
		helper(strS, i, i, &count)
		helper(strS, i, i+1, &count)
	}

	return count
}

func helper(strS []rune, indexStart, indexEnd int, count *int) {
	for indexStart >= 0 && indexEnd < len(strS) && strS[indexStart] == strS[indexEnd] {
		indexStart--
		indexEnd++
		*count++
	}
}
